src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -990,247 +1067,423 @@ RouterSelectTask(Query *originalQuery, Query *query)
+ * provided single shard select task.
+ */
+static Job *
+RouterQueryJob(Query *query, Task *task, List *placementList)
+{
+	Job *job = NULL;
+	List *taskList = NIL;
+	TaskType taskType = task->taskType;
+
 	/*
-	 * We need to make sure there is at least one shard for this hash partitioned
-	 * query to be router plannable. We can not prepare a router plan if there
-	 * are no shards.
+	 * We send modify task to the first replica, otherwise we choose the target shard
+	 * according to task assignment policy. Placement list for select queries are
+	 * provided as function parameter.
 	 */
-	shardCount = ShardIntervalCount(distributedTableId);
-	if (shardCount == 0)
+	if (taskType == MODIFY_TASK)
 	{
-		return false;
+		taskList = FirstReplicaAssignTaskList(list_make1(task));
 	}
+	else
+	{
+		Assert(placementList != NIL);
 
-	return true;
+		task->taskPlacementList = placementList;
+		taskList = list_make1(task);
+	}
+
+	job = CitusMakeNode(Job);
+	job->dependedJobList = NIL;
+	job->jobId = INVALID_JOB_ID;
+	job->subqueryPushdown = false;
+	job->jobQuery = query;
+	job->taskList = taskList;
+
+	return job;
 }
 
 
 /*
- * ColumnMatchExpressionAtTopLevelConjunction returns true if the query contains an exact
- * match (equal) expression on the provided column. The function returns true only
- * if the match expression has an AND relation with the rest of the expression tree.
+ * MultiRouterPlannableQuery returns true if given query can be router plannable.
+ * The query is router plannable if it is a select query issued on a hash
+ * partitioned distributed table, and it has a exact match comparison on the
+ * partition column. This feature is enabled if task executor is set to real-time
  */
-static bool
-ColumnMatchExpressionAtTopLevelConjunction(Node *node, Var *column)
+bool
+MultiRouterPlannableQuery(Query *query, MultiExecutorType taskExecutorType,
+						  RelationRestrictionContext *restrictionContext)
 {
-	if (node == NULL)
+	CmdType commandType = query->commandType;
+	ListCell *relationRestrictionContextCell = NULL;
+
+	if (commandType == CMD_INSERT || commandType == CMD_UPDATE ||
+		commandType == CMD_DELETE)
 	{
-		return false;
+		return true;
 	}
 
-	if (IsA(node, OpExpr))
+	/* FIXME: I tend to think it's time to remove this */
+	if (taskExecutorType != MULTI_EXECUTOR_REAL_TIME)
 	{
-		OpExpr *opExpr = (OpExpr *) node;
-		bool simpleExpression = SimpleOpExpression((Expr *) opExpr);
-		bool columnInExpr = false;
-		bool usingEqualityOperator = false;
-
-		if (!simpleExpression)
-		{
-			return false;
-		}
-
-		columnInExpr = OpExpressionContainsColumn(opExpr, column);
-		if (!columnInExpr)
-		{
-			return false;
-		}
+		return false;
+	}
 
-		usingEqualityOperator = OperatorImplementsEquality(opExpr->opno);
+	Assert(commandType == CMD_SELECT);
 
-		return usingEqualityOperator;
+	if (query->hasForUpdate)
+	{
+		return false;
 	}
-	else if (IsA(node, BoolExpr))
+
+	if (query->hasRecursive)
```

Comment:
On 2016-07-26 11:21:42 -0700, Murat Tuncer wrote:
> >  	}
> > -	else if (IsA(node, BoolExpr))
> > +
> > +	if (query->hasRecursive)
> 
> I could not come up with a good tests for this. I could not comprehend how deep it can go and cause trouble.

Please point out if you re-add restrictions and such, especially when
added without a comment explaining the restriction...  I don't see any
problems with supporting RECURSIVE. Do you need help writing a test case?